Support --disable-tcg on ARM

sameo opened the pull request:
This patchset reorganizes the `target/arm` directory into TCG dependent and independent pieces so that one can build QEMU ARM binaries with `--disable-tcg`.

Commit:
target: arm: Move all v7m insn helpers into their own file

In preparation for supporting TCG disablement on ARM, we move all TCG
related v7m helpers and APIs into their own file (m_helper.c for all
v*-m helpers).

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>
Signed-off-by: Philippe Mathieu-Daudé <philmd@redhat.com>
Tested-by: Philippe Mathieu-Daudé <philmd@redhat.com>
Reviewed-by: Robert Bradford <robert.bradford@intel.com>

Commit:
target: arm: Move all interrupt and exception handlers into their own file

Most of them are TCG dependent so we want to be able to not build them
in order to support TCG disablement with ARM.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>
Tested-by: Philippe Mathieu-Daudé <philmd@redhat.com>
Reviewed-by: Robert Bradford <robert.bradford@intel.com>

Commit:
target: arm: Move the DC ZVA helper into op_helper

Those helpers are a software implementation of the ARM v8 memory zeroing
op code. They should be moved to the op helper file, which is going to
eventually be built only when TCG is enabled.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>
Reviewed-by: Philippe Mathieu-Daudé <philmd@redhat.com>
Tested-by: Philippe Mathieu-Daudé <philmd@redhat.com>
Reviewed-by: Robert Bradford <robert.bradford@intel.com>

Commit:
target: arm: Make ARM TLB filling routine static

It's only used in op_helper.c, it does not need to be exported and
moreover it should only be build when TCG is enabled.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>
Reviewed-by: Philippe Mathieu-Daudé <philmd@redhat.com>
Tested-by: Philippe Mathieu-Daudé <philmd@redhat.com>
Reviewed-by: Robert Bradford <robert.bradford@intel.com>

Commit:
target: arm: Remove the LDST headers

After moving all TCG dependent code out of helper.c,
the software MMU headers are no longer needed.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>
Reviewed-by: Philippe Mathieu-Daudé <philmd@redhat.com>
Tested-by: Philippe Mathieu-Daudé <philmd@redhat.com>
Reviewed-by: Robert Bradford <robert.bradford@intel.com>

Commit:
target: arm: Move all VFP helpers into their own file

Since softfloat is only relevant with TCG, we move all ARM VFP helpers
into their own file (vfp_helper.c), in order to support TCG disablement
on ARM.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>
Signed-off-by: Philippe Mathieu-Daudé <philmd@redhat.com>
Reviewed-by: Philippe Mathieu-Daudé <philmd@redhat.com>
Tested-by: Philippe Mathieu-Daudé <philmd@redhat.com>
Reviewed-by: Robert Bradford <robert.bradford@intel.com>

Commit:
target: arm: Move CPU state dumping routines to helper.c

They're not TCG specific and should be living the generic helper file
instead.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>
Reviewed-by: Robert Bradford <robert.bradford@intel.com>

Commit:
target: arm: Move watchpoints APIs to helper.c

Here again, those APIs are not TCG dependent and can move to the always
built helper.c.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>
Reviewed-by: Philippe Mathieu-Daudé <philmd@redhat.com>
Tested-by: Philippe Mathieu-Daudé <philmd@redhat.com>
Reviewed-by: Robert Bradford <robert.bradford@intel.com>

Commit:
target: arm: Define TCG dependent functions when TCG is enabled

do_interrupt, do_unaligned_access, do_transaction_failed and debug_excp
are only relevant in the TCG context, so we should not define them
when TCG is disabled.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>
Reviewed-by: Philippe Mathieu-Daudé <philmd@redhat.com>
Tested-by: Philippe Mathieu-Daudé <philmd@redhat.com>
Reviewed-by: Robert Bradford <robert.bradford@intel.com>

Commit:
target: arm: Makefile cleanup

Group objects with the same build dependencies together.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>
Reviewed-by: Philippe Mathieu-Daudé <philmd@redhat.com>
Tested-by: Philippe Mathieu-Daudé <philmd@redhat.com>
Reviewed-by: Robert Bradford <robert.bradford@intel.com>

Commit:
target: arm: Do not build TCG objects when TCG is off

We can now safely turn all TCG dependent build off when CONFIG_TCG is
off. This allows building ARM binaries with --disable-tcg.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>
Reviewed-by: Philippe Mathieu-Daudé <philmd@redhat.com>
Tested-by: Philippe Mathieu-Daudé <philmd@redhat.com>
Reviewed-by: Robert Bradford <robert.bradford@intel.com>

sameo commented:
This is the next patchset that I'm planning to send for upstream submission.
This is a split of ef9e21282e0f4f33b1b1c55eabb6cfe97f8539c6 in order to make it potentially upstream acceptable and review-able. This patch is one of the most painful one to rebase for us, so the sooner it makes it upstream, the better for us.

philmd committed:
target: arm: Add copyright boilerplate

Signed-off-by: Philippe Mathieu-Daudé <philmd@redhat.com>
Reviewed-by: Robert Bradford <robert.bradford@intel.com>
Reviewed-by: Samuel Ortiz <sameo@linux.intel.com>

philmd commented:
I wanted to review patch per patch on github but I failed at understanding how to do it (is it possible?).
So I did a local old school review, and pushed (with R-b, T-b) here [philmd/review_nemu_arm_tcg_disable](https://github.com/philmd/qemu/tree/review_nemu_arm_tcg_disable).

```sh
$ git backport-diff -u nemu/topic/upstream/arm-tcg-disable -r 160e5c22e55b3f775c2003dfc626fa872ee4a7a1..
Key:
[----] : patches are identical
[####] : number of functional differences between upstream/downstream patch
[down] : patch is downstream-only
The flags [FC] indicate (F)unctional and (C)ontextual differences, respectively

001/17:[down] 'target/arm: Add copyright boilerplate'
002/17:[down] 'target/arm: Remove unused headers'
003/17:[----] [--] 'target: arm: Move all v7m helpers into their own file'
004/17:[down] '[squash] (c)'
005/17:[down] '[squash] not used'
006/17:[----] [--] 'target: arm: Move all interrupt and exception handlers into their own file'
007/17:[----] [--] 'target: arm: Move the DC ZVA helper into op_helper'
008/17:[----] [--] 'target: arm: Make ARM TLB filling routine static'
009/17:[0001] [FC] 'target: arm: Remove the LDST headers'
010/17:[----] [--] 'target: arm: Move all VFP helpers into their own file'
011/17:[down] '[squash] (c)'
012/17:[down] '[squash] not used'
013/17:[----] [--] 'target: arm: Move CPU state dumping routines to helper.c'
014/17:[----] [--] 'target: arm: Move watchpoints APIs to helper.c'
015/17:[----] [--] 'target: arm: Only define TCG dependent vfuncs when TCG is enabled'
016/17:[----] [--] 'target: arm: Makefile cleanup'
017/17:[----] [--] 'target: arm: Do not build TCG objects when TCG is off'
```

philmd commented:
@sameo in my two `[squash] (c)` commits I misswrote:
```diff
- * This code is licensed under the GNU GPL v2.
+ * This code is licensed under the GNU GPLv2 and later.
```
If you squash them, please fix this! Thanks :)

sameo commented:
@philmd Thanks a lot for the review, I addressed all your comments.

rbradford reviewed (approved):
@sameo great work on splitting this up nicely.

sameo commented:
@rbradford Thanks for the review.
@philmd suggested that I added "Reviewed-by" tags to the patches to those who approve the PR on github. Are you ok with me adding a `Reviewed-by: Rober Bradford <robert.bradford@intel.com>" tag to those patches?

rbradford commented:
@sameo Ack. (If you spell my name correctly :-)

sboeuf commented on target/arm/helper.c:
```diff
@@ -8917,90 +8917,6 @@ hwaddr arm_cpu_get_phys_page_attrs_debug(CPUState *cs, vaddr addr,
 
 #endif
 
-void HELPER(dc_zva)(CPUARMState *env, uint64_t vaddr_in)
```
It'd be nice to have a commit message here, and not only a commit title ;)

sboeuf commented on target/arm/helper.c:
```diff
@@ -16,7 +16,6 @@
 #include "sysemu/sysemu.h"
 #include "qemu/crc32c.h"
 #include "exec/exec-all.h"
-#include "arm_ldst.h"
```
Why is this header no longer needed? Is that because of the re-organization that you applied in the previous commits?
Might be worth explaining why those are no longer needed from the commit message.

sboeuf reviewed (commented):
@sameo this patchset is very nice, very well structured, which makes it easy to review.
The only comment I have is to make sure you provide the right explanation in your commit messages about what you did and why (missing for two commits). Once those have been updated, this PR LGTM!

sameo commented:
v1:
https://lists.gnu.org/archive/html/qemu-devel/2018-11/msg02451.html

sboeuf reviewed (approved):
Thanks for updating the commits messages!
LGTM

Commit:
target: arm: Move v7m exception handling routines to m_helper

In preparation for supporting TCG disablement on ARM, we move the v7m
exception handling routines to m_helper.c
arm_v7m_cpu_do_interrupt pulls a large number of static functions
out of helper.c into m_helper.c because it is TCG dependent.

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

Commit:
target: arm: Interrupt injection

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

rbradford commented:
This project is no longer active and being archived. Cloud Hypervisor is the successor: https://github.com/cloud-hypervisor/cloud-hypervisor
